lib/HeroBonus.cpp

```diff
@@ -948,7 +948,7 @@ void CBonusSystemNode::getAllBonusesRec(BonusList &out) const
 	getAllParents(lparents);
 
 	for(auto parent : lparents)
-		parent->bonuses.getAllBonuses(beforeUpdate);
+		parent->getAllBonusesRec(beforeUpdate);

```

Comment:
Does getAllParents return direct parents or grandparents as well? Looks like all of them and parent->getAllBonusesRec will do the same so it looks like grandparent will be iterated twice and this might cause same bonus to be activated a few times which might be wrong for additive bonuses.

Reply:
That's how it worked exactly before the changes from Sep 2021. I had same concern, but looks like duplicated bonuses will be discarded later in limiters as it will be processed for improper parent.
I've tested it a lot and didn't discover issues (but I'm bad in QA:)) do you have scenarios in mind which need to be verified?

Reply:
It's not the ideal but it fixes regression (as for me missing specialities is a blocker for the release)